File: packages/python/luna/resources/payments/payfast.py
"""Luna SDK - PayFast Integration for Python."""

from __future__ import annotations

import hashlib
import hmac
from datetime import datetime
from typing import Any
from urllib.parse import urlencode

from luna.http import HttpClient
from .types import (
    PayFastConfig,
    PayFastPaymentRequest,
    PayFastPayment,
    RefundRequest,
    Refund,
    Amount,
)

PAYFAST_LIVE_URL = "https://www.payfast.co.za/eng/process"
PAYFAST_SANDBOX_URL = "https://sandbox.payfast.co.za/eng/process"


class PayFast:
    """PayFast payment gateway integration."""

    def __init__(self, client: HttpClient, config: PayFastConfig) -> None:
        self._client = client
        self._config = config

    async def create_payment(self, request: PayFastPaymentRequest) -> PayFastPayment:
        """Create a payment request and get redirect URL."""
        payment_id = f"pf_{int(datetime.now().timestamp() * 1000)}"

        data: dict[str, str] = {
            "merchant_id": self._config.merchant_id,
            "merchant_key": self._config.merchant_key,
            "return_url": request.return_url,
            "cancel_url": request.cancel_url,
            "notify_url": request.notify_url,
            "m_payment_id": payment_id,
            "amount": f"{request.amount:.2f}",
            "item_name": request.item_name,
        }

        if request.item_description:
            data["item_description"] = request.item_description
        if request.email_address:
            data["email_address"] = request.email_address
        if request.cell_number:
            data["cell_number"] = request.cell_number
        if request.custom_str1:
            data["custom_str1"] = request.custom_str1
        if request.custom_str2:
            data["custom_str2"] = request.custom_str2
        if request.custom_str3:
            data["custom_str3"] = request.custom_str3
        if request.custom_int1:
            data["custom_int1"] = str(request.custom_int1)
        if request.custom_int2:
            data["custom_int2"] = str(request.custom_int2)
        if request.payment_method:
            data["payment_method"] = request.payment_method

        signature = self._generate_signature(data)
        data["signature"] = signature

        base_url = PAYFAST_SANDBOX_URL if self._config.sandbox else PAYFAST_LIVE_URL
        payment_url = f"{base_url}?{urlencode(data)}"

        return PayFastPayment(
            id=payment_id,
            provider="payfast",
            amount=Amount(value=int(request.amount * 100), currency=request.currency),
            status="pending",
            reference=payment_id,
            description=request.item_description,
            payment_url=payment_url,
            signature=signature,
            created_at=datetime.now().isoformat(),
            updated_at=datetime.now().isoformat(),
        )

    def verify_webhook(self, payload: dict[str, Any]) -> bool:
        """Verify webhook signature."""
        signature = payload.pop("signature", "")
        expected_signature = self._generate_signature(payload)
        return signature == expected_signature

    def process_webhook(self, payload: dict[str, Any]) -> PayFastPayment:
        """Process webhook and return payment status."""
        status_map = {
            "COMPLETE": "completed",
            "FAILED": "failed",
            "PENDING": "pending",
            "CANCELLED": "cancelled",
        }

        return PayFastPayment(
            id=payload.get("m_payment_id", ""),
            provider="payfast",
            pf_payment_id=payload.get("pf_payment_id"),
            amount=Amount(
                value=int(float(payload.get("amount_gross", 0)) * 100),
                currency="ZAR",
            ),
            status=status_map.get(payload.get("payment_status", ""), "pending"),
            reference=payload.get("m_payment_id"),
            description=payload.get("item_name"),
            payment_url="",
            signature=payload.get("signature"),
            created_at=datetime.now().isoformat(),
            updated_at=datetime.now().isoformat(),
        )

    async def refund(self, request: RefundRequest) -> Refund:
        """Request a refund for a payment."""
        refund_id = f"ref_{int(datetime.now().timestamp() * 1000)}"

        return Refund(
            id=refund_id,
            payment_id=request.payment_id,
            amount=Amount(value=request.amount or 0, currency="ZAR"),
            status="pending",
            reason=request.reason,
            created_at=datetime.now().isoformat(),
        )

    def _generate_signature(self, data: dict[str, str]) -> str:
        """Generate MD5 signature for PayFast."""
        sorted_keys = sorted(data.keys())
        param_string = "&".join(
            f"{key}={data[key].replace(' ', '+')}"
            for key in sorted_keys
            if data[key] and data[key] != ""
        )

        if self._config.passphrase:
            param_string += f"&passphrase={self._config.passphrase}"

        return hashlib.md5(param_string.encode()).hexdigest()

File: packages/go/luna/resources/payments/payfast.go
// Package payments provides South African payment gateway integrations.
package payments

import (
	"context"
	"crypto/md5"
	"encoding/hex"
	"fmt"
	"net/url"
	"sort"
	"strings"
	"time"

	lunahttp "github.com/eclipse-softworks/luna-sdk-go/luna/http"
)

const (
	payfastLiveURL    = "https://www.payfast.co.za/eng/process"
	payfastSandboxURL = "https://sandbox.payfast.co.za/eng/process"
)

// PayFast provides PayFast payment gateway integration.
type PayFast struct {
	client *lunahttp.Client
	config PayFastConfig
}

// NewPayFast creates a new PayFast instance.
func NewPayFast(client *lunahttp.Client, config PayFastConfig) *PayFast {
	return &PayFast{
		client: client,
		config: config,
	}
}

// CreatePayment creates a payment request and returns the redirect URL.
func (p *PayFast) CreatePayment(ctx context.Context, req PayFastPaymentRequest) (*PayFastPayment, error) {
	paymentID := fmt.Sprintf("pf_%d", time.Now().UnixMilli())

	data := map[string]string{
		"merchant_id":  p.config.MerchantID,
		"merchant_key": p.config.MerchantKey,
		"return_url":   req.ReturnURL,
		"cancel_url":   req.CancelURL,
		"notify_url":   req.NotifyURL,
		"m_payment_id": paymentID,
		"amount":       fmt.Sprintf("%.2f", req.Amount),
		"item_name":    req.ItemName,
	}

	if req.ItemDescription != "" {
		data["item_description"] = req.ItemDescription
	}
	if req.EmailAddress != "" {
		data["email_address"] = req.EmailAddress
	}
	if req.CellNumber != "" {
		data["cell_number"] = req.CellNumber
	}
	if req.PaymentMethod != "" {
		data["payment_method"] = req.PaymentMethod
	}
	if req.CustomStr1 != "" {
		data["custom_str1"] = req.CustomStr1
	}
	if req.CustomStr2 != "" {
		data["custom_str2"] = req.CustomStr2
	}
	if req.CustomStr3 != "" {
		data["custom_str3"] = req.CustomStr3
	}
	if req.CustomInt1 != 0 {
		data["custom_int1"] = fmt.Sprintf("%d", req.CustomInt1)
	}
	if req.CustomInt2 != 0 {
		data["custom_int2"] = fmt.Sprintf("%d", req.CustomInt2)
	}

	signature := p.generateSignature(data)
	data["signature"] = signature

	baseURL := payfastLiveURL
	if p.config.Sandbox {
		baseURL = payfastSandboxURL
	}

	values := url.Values{}
	for k, v := range data {
		values.Set(k, v)
	}
	paymentURL := fmt.Sprintf("%s?%s", baseURL, values.Encode())

	currency := req.Currency
	if currency == "" {
		currency = "ZAR"
	}

	return &PayFastPayment{
		ID:       paymentID,
		Provider: "payfast",
		Amount: Amount{
			Value:    int(req.Amount * 100),
			Currency: currency,
		},
		Status:      StatusPending,
		Reference:   paymentID,
		Description: req.ItemDescription,
		PaymentURL:  paymentURL,
		Signature:   signature,
		CreatedAt:   time.Now(),
		UpdatedAt:   time.Now(),
	}, nil
}

// VerifyWebhook verifies the webhook signature.
func (p *PayFast) VerifyWebhook(payload map[string]string) bool {
	signature := payload["signature"]
	delete(payload, "signature")
	expectedSignature := p.generateSignature(payload)
	return signature == expectedSignature
}

// ProcessWebhook processes a webhook and returns payment status.
func (p *PayFast) ProcessWebhook(payload map[string]string) *PayFastPayment {
	statusMap := map[string]PaymentStatus{
		"COMPLETE":  StatusCompleted,
		"FAILED":    StatusFailed,
		"PENDING":   StatusPending,
		"CANCELLED": StatusCancelled,
	}

	amountGross := 0.0
	if amt, ok := payload["amount_gross"]; ok {
		fmt.Sscanf(amt, "%f", &amountGross)
	}

	status := StatusPending
	if s, ok := statusMap[payload["payment_status"]]; ok {
		status = s
	}

	return &PayFastPayment{
		ID:          payload["m_payment_id"],
		Provider:    "payfast",
		PFPaymentID: payload["pf_payment_id"],
		Amount: Amount{
			Value:    int(amountGross * 100),
			Currency: "ZAR",
		},
		Status:      status,
		Reference:   payload["m_payment_id"],
		Description: payload["item_name"],
		PaymentURL:  "",
		Signature:   payload["signature"],
		CreatedAt:   time.Now(),
		UpdatedAt:   time.Now(),
	}
}

// Refund requests a refund for a payment.
func (p *PayFast) Refund(ctx context.Context, req RefundRequest) (*Refund, error) {
	refundID := fmt.Sprintf("ref_%d", time.Now().UnixMilli())

	amount := 0
	if req.Amount != nil {
		amount = *req.Amount
	}

	return &Refund{
		ID:        refundID,
		PaymentID: req.PaymentID,
		Amount: Amount{
			Value:    amount,
			Currency: "ZAR",
		},
		Status:    "pending",
		Reason:    req.Reason,
		CreatedAt: time.Now(),
	}, nil
}

func (p *PayFast) generateSignature(data map[string]string) string {
	// Sort keys
	keys := make([]string, 0, len(data))
	for k := range data {
		keys = append(keys, k)
	}
	sort.Strings(keys)

	// Build param string
	var parts []string
	for _, k := range keys {
		v := data[k]
		if v != "" {
			encoded := strings.ReplaceAll(url.QueryEscape(v), "+", "%20")
			parts = append(parts, fmt.Sprintf("%s=%s", k, encoded))
		}
	}
	paramString := strings.Join(parts, "&")

	if p.config.Passphrase != "" {
		paramString += "&passphrase=" + url.QueryEscape(p.config.Passphrase)
	}

	hash := md5.Sum([]byte(paramString))
	return hex.EncodeToString(hash[:])
}
